### PR TITLE
Float(n) != Rational(n)

### LINKED ISSUE #11707: a == b doesn't imply hash(a) == hash(b) for floats
See https://github.com/sympy/sympy/pull/10657#r53846283.

```
In [17]: S('0.3') == Rational(3, 10)
Out[17]: True

In [18]: hash(S('0.3')) == hash(Rational(3, 10))
Out[18]: False
```

I've written a blog post about things that can go wrong here (https://asmeurer.github.io/blog/posts/what-happens-when-you-mess-with-hashing-in-python/). You can already see, for instance

```
In [19]: [S('0.3')] == [Rational(3, 10)]
Out[19]: True

In [20]: {S('0.3')} == {Rational(3, 10)}
Out[20]: False

In [22]: S("0.3") in [Rational(3, 10)]
Out[22]: True

In [23]: S("0.3") in {Rational(3, 10)}
Out[23]: False
```

I don't know how to fix this.

### LINKED ISSUE #19915: Structural equality for numbers (Rational and Float)
I'm trying to understand what's triggering [this issue](https://github.com/sympy/sympy/issues/19774), which could be related to [this question](https://groups.google.com/g/sympy/c/8uSZNb6-z54).

Following the trace, I realized that `Rational` and `Float` are not implementing a proper structural equality by overloading `__eq__`. In fact:

```
(S(5) / 2) == S(2.5)
# True
```

### LINKED ISSUE #19976: Float with replace
`(2*x+7).evalf().replace(lambda e: e.is_Float, lambda e: Integer(e))` doesn't give `2x+7`

### PR DESCRIPTION
**Pull request by smichr**
continuation of #20033 

#### References to other Issues or PRs

Follows #16924
Fixes #19915 
Fixes #19976 
Fixes #11707
(Probably other issues)

#### Brief description of what is fixed or changed

Previously @smichr's work in #16924 made it so that expressions containing Float would compare unequal to expressions containing Rationals e.g.:
```julia
In [1]: e1 = x**2

In [2]: e2 = x**2.0

In [3]: e1
Out[3]: 
 2
x 

In [4]: e2
Out[4]: 
 2.0
x   

In [5]: e1 == e2
Out[5]: False
```
However #16924 left the behaviour unchanged for individual Number comparisons e.g.:
```julia
In [6]: Float(2.0) == Integer(2)
Out[6]: True
```
This PR extends that change to apply to individual Number so with the PR we have:
```julia
In [1]: Float(2.0) == Integer(2)
Out[1]: False
```
Moreover a Float will only compare equal to another Float if they have the same precision:
```julia
In [3]: Float(2.0, 5) == Float(2.0, 6)
Out[3]: False

In [4]: Float(2.0, 5) == Float(2.0, 5)
Out[4]: True
```
Only a `Float` created with default precision will compare equal to a Python `float`:
```julia
In [5]: Float(2.0, 5) == 2.0
Out[5]: False

In [6]: Float(2.0) == 2.0
Out[6]: True
``` 
The hash function for Float is changed to return the same as `float.__hash__` if the Float can be represented as a float (e.g. if the exponent is not too large/small):
```julia
In [7]: hash(2.3)
Out[7]: 691752902764107778

In [8]: hash(Float(2.3))
Out[8]: 691752902764107778
```

On master expressions involving Float compare equal but with distinct hashes:
```julia
In [1]: e1 = x**Float(5, 10)

In [2]: e2 = x**Float(5, 20)

In [3]: e1
Out[3]: 
 5.0
x   

In [4]: e2
Out[4]: 
 5.0
x   

In [5]: e1 == e2
Out[5]: True

In [6]: hash(e1)
Out[6]: 1397520323187279713

In [7]: hash(e2)
Out[7]: 7223358913859886252
```
With this PR we get
```julia
In [5]: e1 == e2
Out[5]: False

In [6]: hash(e1)
Out[6]: 5346466018815759790

In [7]: hash(e2)
Out[7]: 5346466018815759790
```
It is not necessary that the hash functions be equal in this example. That's just a consequence of ensuring that they are equal between Float and float. The important thing is that the expressions compare unequal. In that situation there are no constraints on whether the hashes should be equal or not.

This fixes #11707 by ensuring that Floats respect the hash invariant:
```
a == b   -->   hash(a) == hash(b)
```
This also ensures that comparisons using `==` with Floats respect structural equality. Expressions that differ in use of Float/Rational are not structurally equal and the same goes for Floats with differing precision. In this PR those will no longer compare equal.

### cps comments
To ignore differences in precision (or to compare numbers in an arbitrary way in an expression) the `is_same` method can be used with a custom function for comparing numbers:
```python
>>> from sympy.core.numbers import equal_valued
>>> e1.is_same(e2)
False
>>> e1.is_same(e2, equal_valued)
True
```
The `equal_valued` function is strict in terms of checking that the internal base-2 representation of the number is the same while ignoring the precision. This means that floats with a decimal portion that is not representable as a fraction with a denominator that is a power of 2 will not compare the same.
```python
>>> equal_valued(Float(.5), Float(.5, 20))
True
>>> equal_valued(Float(.1), Float(.1, 20))
False
```
There are other options for testing such numbers (and any function that can do so can be passed to `is_same`), see, for example [this description and comparison to similar functions](https://docs.pytest.org/en/latest/reference/reference.html#pytest-approx)
```python
>>> from sympy import root
>>> import math
>>> root(x,10).is_same(x**Float(.1), equal_valued)
False
>>> root(x,10).is_same(x**Float(.1), math.isclose)
True
```
#### Other comments

This is not yet finished. In particular after making various changes to fix tests I have found that there seems to be a need for some sort of `equalf` function that would compare expressions while ignoring the distinction between Float/Rational or Floats of differing precision.

We could have a function `equalf` so that e.g.:
```
>>> equalf(Float(2.0), 2)
True
>>> equalf(S.Half, 0.5)
True
>>> equalf(x**2.0, x**2)
True
```
#### UPDATE: this can be done by passing the desired number-comparing function like `equal_valued` to `is_same`.

There probably also needs to be a function for determining whether or not a Float/Rational represents an integer value. Currently around the codebase we have many cases that do `int(a) == a` which fails in this PR. Instead I have replaced those with `a.round(0) == a` since `a.round(0)` returns a `Float`. With `equalf` you could do `equalf(a, int(a))` but that could be problematic if `a` has a large exponent.

#### UPDATE: this can be done the same way we test for odd or even numbers: to check for fraction on `n` use `n % 1` but there is a faster test added to numbers.py named `int_valued` that returns True when passed a number that has no decimal part.

Clearly this change is not backwards compatible. I expect that many users will find this more confusing than the current situation but I think it will help with sympy's internal consistency and will reduce bugs in future. For example the bug in #19976 is fixed here:
```julia
In [16]: (2.0*x).replace(lambda e: e.is_Float, Integer)
Out[16]: 2⋅x
```
On master that gives `2.0*x`. I have fixed many other bugs that were ultimately caused by Float comparisons in the past.

One particular point of confusion is that in this PR sympy's Float/Integer no longer behave analogously to Python's float/int:
```julia
In [11]: a, b = 2, 2.0

In [12]: a == b
Out[12]: True

In [13]: S(a) == S(b)
Out[13]: False
```
This also breaks transitivity of `==` although I don't think that's actually a problem any of Python's core data structures in the way that breaking the hash invariant is:
```julia
In [14]: Integer(1) == 1 == 1.0 == Float(1.0) != Integer(1)
Out[14]: True
```

I'm putting this up to discuss whether this is wanted and consider whether it is worth a significant break in backwards compatibility.

### cps
It is past-due to have this change added since it is hard to defend how `sqrt(x) != x**.5` while `S.Half == .5` and `x == x`.

There are semantic issues to consider now, namely: what type of comparison should be used for internal simplification
```python

>>> for f in (Eq, equal_valued, Basic.__eq__):
...    print(f(S(1), S(1.)), f(S.Half, S(.5)), f(Rational(1,10), Float(.1)))
...
True True True
True True False
False False False

>>> Min(1, 1.0)
1.0
>>> Max(1, 1.0)
1

>>> And(x<1,x<1.)
(x < 1.0) & (x < 1)
```
#### Release Notes

* core
    * BREAKING CHANGE: Float and Integer/Rational no longer compare equal with `a == b`. From now on `Float(2.0) != Integer(2)`. Previously expressions involving `Float` would compare unequal e.g. `x*2.0 != x*2` but an individual `Float` would compare equal to an `Integer`. In SymPy 1.7 a `Float` will always compare unequal to an `Integer` even if they have the same "value". Use `sympy.numbers.int_valued(number)` to test if a number is a concrete number with no decimal part.
    * BREAKING CHANGE: Floats with different prevision no longer compare equal with `a == b`. From now on `Float(2.0, 10) != Float(2.0, 5)`. Previously expressions involving `Float` with different precision would compare equal although they would have different hashes. Now expressions compared with `a == b` involving Floats respect structural equality since Floats of differing precision are not interchangeable.
    * The hash function for Floats and expressions involving Float now respects the hash invariant that if `a == b` then `hash(a) == hash(b)`. This ensures that it is possible to use such expressions in a guaranteed deterministic way in Python's core `set` and `dict` data structures. It is however not recommended to mix up sympy `Basic` objects with non-sympy number types such as core Python's `float` or `int` in the same `set/dict`.
    * It is now possible to replace `Float` with `Integer` in an expression using the `Basic.replace` method.
    * Basic.is_same(a, b) returns True only if all classes are exactly the same; has option to test numbers with custom rule

### CODE REVIEW

**Comment by jksuom** [comment]
> A problem that still exists is that Float(0)*1 -> 0 so even though Float(0) != 0, Float(0)*1 == 0*1.

Float(0) is unique among floating point numbers: it is not inexact. We could make Float(0) return S(0) in the same way as Float('inf') returns oo.

**Comment by smichr** [comment]
> Float(0) is unique among floating point numbers: it is not inexact.

This must be a semantic interpretation. There are two ways to interpret the inexactness of a Float:
1. the hardware limitation of representing the float as a fraction with power-of-two denominator
2. the inherent limited precision of the number

For case 2, I imagine the numbers has being `x +/- dev` and if we subtract to numbers that have the same x we don't get 0, we get `0 +/- dev'`.

If we are going to interpret Float as a convenience of representing numbers in an inexact way (case 1) then allowing `Float(0) -> S(0)` will potentially shift you out of a desired computational theme (Floats) to Rational. And just as shifting from int to Integer in some routines is bad for performance, shifting from Float to Rational would also be bad. BUT...if all other numbers in that routine are Float your state of being Rational will last only as long as it is until he next operation that adds a Float.

So under interpretation (1) I agree with the idea of letting `Float(0) -> S(0)` as long as this doesn't lose the ability to compute a "precisionless 0":
```python
>>> (cos(1)**2+sin(1)**2-1).n()
-0.e-124
>>> _._prec
1
```

**Comment by ThePauliPrinciple** [comment]
An alternative option to avoid possible equality/hash issue is to say that python `float` and sympy `Float` do not compare equal. In general, I do not think they are the same and I would encourage anyone to use a `Rational` if in python they'd use a float before attempting to use a `Float`. 

I'm not sure how practical it is w.r.t. to `float == Float` comparisons currently in sympy.

**Comment by smichr** [comment]
Check the changes to the printing tests, esp. fortran: is it a good idea to no longer be able to represent `Float(0)` as 0.0?

**Comment by smichr** [comment]
`S.Zero` has to have the `_mpf_ = (0,0,0,0)` property to it will be recognized by mpmath.

File: sympy/printing/tests/test_latex.py [hunk 1]
```diff
@@ -121,7 +121,7 @@ def test_latex_basic():
     assert latex(2*x*y) == r"2 x y"
     assert latex(2*x*y, mul_symbol='dot') == r"2 \cdot x \cdot y"
     assert latex(3*x**2*y, mul_symbol='\\,') == r"3\,x^{2}\,y"
-    assert latex(1.5*3**x, mul_symbol='\\,') == r"1.5 \cdot 3^{x}"
+    assert latex(1.5*3**x, mul_symbol='\\,') == r"3^{x}\,1.5"
```
Comment (smichr) [inline sympy/printing/tests/test_latex.py:124]: This is a mystery to me: I have no idea why the previous test passed.

File: sympy/core/tests/test_expr.py [hunk 2]
```diff
@@ -2072,7 +2077,7 @@ def test_round():
         assert all(type(round(i, p)) is int for p in (-1, 0, 1))
         assert all(S(i).round(p).is_Integer for p in (-1, 0, 1))
         assert all(type(round(fi, p)) is float for p in (-1, 0, 1))
-        assert all(S(fi).round(p).is_Float for p in (-1, 0, 1))
+        assert all(S(fi).round(p).is_Float or not i for p in (-1, 0, 1))
```
Comment (smichr) [inline sympy/core/tests/test_expr.py:2080]:
```suggestion
        assert all(S(fi).round(p).is_Float or not i or p == -1 for p in (-1, 0, 1))
```

**Comment by oscarbenjamin** [comment]
I think it would be better to try to break this down rather than doing everything in one PR as I suggested in https://github.com/sympy/sympy/pull/20033. I took some of the changes out in #24625 but the aim should be to reduce the diff by making non-controversial changes in the first instance.

I don't think that we are ready to actually make the main change until there are helpers that can make it easy to manage the change and I don't think that the helpers should be added in the same release as the main change: we need to add the helpers, let them sit for a while, and then when it is clear that the changes needed here can be made easily we can consider actually making the change.

**Comment by smichr** [comment]
> I don't think that we are ready to actually make the main change until there are helpers that can make it easy to manage the change

I'm not sure of what more you are thinking. 

In master, Floats of different precision don't compare the same unless they represent fractions with a power of 2 denominator). If one wants the same behavior as master when comparing numbers they can be compared with `equal_valued`. But for the special case of detecting when a Float has no decimal portion, instead of using `equal_valued(Float(3), 3)` the user can check `not Float(3)%1`: there is no need for a new function since the user can test this just like testing for whether a number is odd or even.

File: sympy/core/numbers.py [hunk 3]
```diff
@@ -778,14 +783,16 @@ def __new__(cls, num, dps=None, precision=None):
             elif not _literal_float(num):
                 raise ValueError('string-float not recognized: %s' % _num)
         elif isinstance(num, float) and num == 0:
-            num = '0'
+            return S.Zero  # there is not Float(0)
```
Comment (oscarbenjamin) [inline sympy/core/numbers.py:786]: I am not sure about removing `Float(0)`.

**Comment by oscarbenjamin** [comment]
> But for the special case of detecting when a Float has no decimal portion, instead of using `equal_valued(Float(3), 3)` the user can check `not Float(3)%1`: there is no need for a new function

It is better to have a special function. With symbolic expressions `expr % 1` creates a new expression which is unnecessary just to check if the original expression is integer valued. A special function that looks at the internal data structure of a Float can determine if it is integer-valued much more efficiently (just compare the bit-length of the mantissa with the exponent). More importantly a function with a good name communicates intent better.

File: sympy/core/numbers.py (hunk 3 above)
Comment (smichr) [reply sympy/core/numbers.py:786]: see https://github.com/sympy/sympy/pull/25614#issuecomment-1703756842

Comment (smichr) [reply sympy/core/numbers.py:786]: Instead of removing it maybe we could all `Float(0) == 0` to be the only exception. Maybe that was @jksuom intent from the start.

Comment (oscarbenjamin) [reply sympy/core/numbers.py:786]: What I don't like is that we lose the distinction between exact and approximate.

**Comment by smichr** [comment]
> just compare the bit-length of the mantissa with the exponent

It was my understanding that mpf for `x` is `(s, m, e, p)` with `x == s*m*2**e`. So isn't `x` and integer if `e >= 0`?

**Comment by oscarbenjamin** [comment]
> So isn't `x` and integer if `e >= 0`?

Yes, I think so.

File: sympy/core/numbers.py (hunk 3 above)
Comment (smichr) [reply sympy/core/numbers.py]: Then we need to make `Float(0)*1 -> Float(0)`. I can undo the `Float(0) == 0` and leave make `Float(0)*1 = Float(0)` for another PR. It, too, is not a trivial change because of the code that depends on this behavior.

File: sympy/series/tests/test_limits.py [hunk 4]
```diff
@@ -421,10 +421,10 @@ def test_issue_4547():
 
 def test_issue_5164():
     assert limit(x**0.5, x, oo) == oo**0.5 is oo
-    assert limit(x**0.5, x, 16) == S(16)**0.5
+    assert limit(x**0.5, x, 16) == 4 # Should this be a float?
     assert limit(x**0.5, x, 0) == 0
     assert limit(x**(-0.5), x, oo) == 0
-    assert limit(x**(-0.5), x, 4) == S(4)**(-0.5)
+    assert limit(x**(-0.5), x, 4) == S.Half # Should this be a float?
```
Comment (oscarbenjamin) [inline sympy/series/tests/test_limits.py:427]: There was an issue and a PR for this at some point...

File: sympy/sets/handlers/intersection.py [hunk 5]
```diff
@@ -435,10 +435,20 @@ def _(a, b):
...
-            #this is to ensure that if Eq(a.start,b.start) but
-            #type(a.start) != type(b.start) the order of a and b
-            #does not matter for the result
-            start = list(ordered([a,b]))[0].start
+            start = a.start
+            if a.start != b.start:
+                # For example Integer(2) != Float(2)
+                # Prefer the Float boundary because Floats should be
+                # contagious in calculations.
+                if b.start.has(Float) and not a.start.has(Float):
```
Comment (oscarbenjamin) [inline sympy/sets/handlers/intersection.py:443]: This change is another good candidate to split out to a separate PR.

File: sympy/series/tests/test_limits.py (hunk 4 above)
Comment (smichr) [reply sympy/series/tests/test_limits.py:427]: https://github.com/sympy/sympy/pull/23539

File: sympy/sets/handlers/intersection.py (hunk 5 above)
Comment (smichr) [reply sympy/sets/handlers/intersection.py:443]: As mentioned in #25662 in which this was placed, this change will not work until this PR is in.

**Comment by smichr** [comment]
Any objections to committing this as is? `Float(0)*1 -> Float(0) instead of S(0)` can be fixed in another PR and is a pre-existing blemish not related to this PR.

File: sympy/core/numbers.py [hunk 6]
```diff
@@ -1101,37 +1100,22 @@ def __int__(self):
...
-        if not self:
-            return not other
-        return False    # Float != non-Number
+        if isinstance(other, float):
+            other = Float(other)
+        return Basic.__eq__(self, other)
 
     def __ne__(self, other):
-        return not self == other
+        eq = self.__eq__(other)
```
Comment (ThePauliPrinciple) [inline sympy/core/numbers.py:1108]: What is the reason for this over the standard __ne__ implementation?

File: sympy/core/tests/test_numbers.py [hunk 7]
```diff
@@ -190,6 +190,7 @@ def test_divmod():
     assert divmod(S(-3), S(2)) == (-2, 1)
     assert divmod(S(-3), 2) == (-2, 1)
 
+
```
Comment (ThePauliPrinciple) [inline sympy/core/tests/test_numbers.py:193]: I think we don't want double newline within a function

File: sympy/core/tests/test_numbers.py [hunk 8]
```diff
@@ -529,8 +532,10 @@ def teq(a):
 
     # inexact floats (repeating binary = denom not multiple of 2)
     # cannot have precision greater than 15
-    assert Float(.125, 22) == .125
-    assert Float(2.0, 22) == 2
+    assert Float(.125, 22)._prec == 76
+    assert Float(2.0, 22)._prec == 76
+    #assert Float(.125, 22) == .125
```
Comment (ThePauliPrinciple) [inline sympy/core/tests/test_numbers.py:537]: I think commented code should have some reason why it is there (in the comments) and I think linters discourage comments without a space right after the #

File: sympy/core/numbers.py (hunk 6 above)
Comment (smichr) [reply sympy/core/numbers.py:1108]: I didn't write this but judging from the code one must handle the NotImplementedError. And rather than duplicate what is done in  `__eq__` we just call `__eq__` and return the the `not` of it.

File: sympy/core/tests/test_numbers.py (hunk 7 above)
Comment (smichr) [inline sympy/core/tests/test_numbers.py:193]:
```suggestion
```

File: sympy/core/tests/test_numbers.py (hunk 8 above)
Comment (smichr) [inline sympy/core/tests/test_numbers.py:537]:
```suggestion
    # only default prec is equal, even for exactly representable float
    assert Float(.125, 22) != .125
```

File: sympy/core/numbers.py (hunk 6 above)
Comment (smichr) [reply sympy/core/numbers.py:1108]: @oscarbenjamin , do you have any feedback for this?

Comment (smichr) [reply sympy/core/numbers.py:1108]: BTW, `not NotImplementedError` is still NotImplementedError so we could return `not self.__eq__(other)` -- is that more what you expect to see, @ThePauliPrinciple ?

**Comment by smichr** [comment]
If `__ne__` needs modification that can be done in a new PR.

**Comment by oscarbenjamin** [comment]
I'm still not sure about this change. I don't think that it should have been merged without discussion.

**Comment by oscarbenjamin** [comment]
I think this has now broken the tests on master.

**Comment by smichr** [comment]
> I'm still not sure about this change. I don't think that it should have been merged without discussion.

I did not sense a strong feeling against this and a thought this was much more in line with `==` tests for structural equality. (Though the `is_same` docstring shows how this is not followed for some objects.) 

`Eq` still works to compare numbers without regard to precision (though that should probably change else `Eq(2, 2.0)` is True while `Eq(x**2, x**2.0)` is unevaluated). There are functions to test whether a number is an integer, `int_valued` and whether number are equal as defined by `equal_valued` or using any other comparison like `comp`, `pytest.approx`, `math.isclose`, (might have parent wrong in those refs) etc... in `is_same`.

Although I can no longer see see why we would allow `S(2) == Float(2)` but not `x**2 == x**2.0`, if you don't think this should be in we can revert it.

**Comment by oscarbenjamin** [comment]
I would just like others to review this change and give their opinions. I know that when I first proposed this it seemed a bit controversial.

I am almost certain that this will cause some downstream breakage.

**Comment by asmeurer** [comment]
Was the Float(0) -> Integer(0) thing that was being discussed above end up being done here? 

Personally I think it's a bad idea. Every integer within the Float precision is represented exactly, so there's no reason why 0 is particularly special. And as it was pointed out, Float(0) carries the additional information of its precision. Mixing floats with different precisions, or even using floats without the default precision isn't very common, so I wouldn't be surprised if there are bugs and design issues there in general, but it should work in principle at least. To contrast, the precision is somewhat meaningless for nan and infinities, so there's less of an issue there with only having one symbolic value. 

0 tends to be treated specially in floating-point contexts, but it's typically more just because it happens to be a special value in equations, and also due to the way the IEEE 754 representation works. I believe that if we actually did this, it would lead to weird bugs where things behaved differently for one specific value because that value resulted in a float 0.0 somewhere instead of float something-else.

**Comment by jksuom** [comment]
> Float(0) carries the additional information of its precision.

Float(0) has the attribute _prec in SymPy (53 by default) but what does it mean? As I understand it, a Float x is generally used to denote a real number  x +/- dev  where dev is bounded by a fraction of the absolute value of x (of the order of `2**-prec). But if x is zero, how does its _prec affect dev? Can it be nonzero, and if so, then what should be its order of magnitude?

**Comment by oscarbenjamin** [comment]
> Float(0) has the attribute _prec in SymPy (53 by default) but what does it mean?

The meaning of `Float` and `_prec` is generally unclear in SymPy. As returned by `evalf` it represents a sort of bound on accuracy/certainty. In the context of arithmetic it effectively represents a statement about how much precision is desired:
```python
In [1]: f = Rational(1, 3).n(5)

In [2]: f
Out[2]: 0.33333

In [3]: f * 0.1  # Uses the larger _prec (implicitly 53 for 0.1)
Out[3]: 0.0333333492279053

In [4]: f = Rational(1, 3).n(50)

In [5]: f
Out[5]: 0.33333333333333333333333333333333333333333333333333

In [6]: f * 0.1
Out[6]: 0.033333333333333335183705041041927567372719446818034

In [7]: f * Rational(1, 10)  # f._prec determines precision for Rational(1, 10).evalf()
Out[7]: 0.033333333333333333333333333333333333333333333333333
```
To me the main point is that I want it to be clear when something is approximate or exact. Any `Float(0)` cannot be understood as meaning exactly zero because it arises from some floating point operation. This is related to:
```python
In [3]: x**1.0
Out[3]: 
 1.0
x  
```
We want to know that this is not exactly just `x` because something approximate happened somewhere.

**Comment by asmeurer** [comment]
I think it's a mistake to think of Float as representing an interval/ball approximation. A true ball arithmetic library like arb would handle this sort of thing properly, but mpmath is not that. 

Rather, I think of a Float a just a fixed rational approximation of some value. In a calculation, the precision is used to give something like an approximate interval to bound that value, but once the calculations are done, the result is just a single value. Any rounding errors that occurred during this computation are sort of baked into the float. Any loss of significance that occured as part of computing the value is already gone. That's really why floats are so bad, because you can't tell if you've lost precision, and multi-precision floats aren't any better. Ball arithmetic is much better in this regard.

(there's a bit of a complication here in that SymPy Floats store their precision on themselves, meaning floats of different precisions can be combined. But this is just an abstraction on top of mpmath, which only supports a single global precision inside of a calculation. So whenever a calculation is done, some global precision is chosen based on the precision(s) of inputs)

The way I think of it is that a float is an approximation of whatever value you care about. At each stage of a computation, you get a new approximation, and some error is (necessarily) introduced. The computation can make a best effort to reduce the error added at that stage, but once the result is returned, that's it. Whatever error was produced is there now, and there's no way to learn anything about it. This mostly works with multi-precision floats because you can typically just increase your working precision to minimize the error at each level of the computation to lower than your result precision (what mpmath does), and also because you can just run the same computation again with a higher precision and compare the output (what evalf does). But at the end of the day, any floating-point value is still just an approximation with absolutely no data on where it came from. 

So when you see a Float, you can use its precision to make assumptions about computations going forward, but you shouldn't make any assumptions about where things came from going backwards. You can only assume that it is "correct" to the number of bits specified by the prec (which is more or less the same as saying it is whatever rational number it represents).
